File: claude-agent-sdk/src/main/kotlin/com/asakii/claude/agent/sdk/types/Permissions.kt
package com.asakii.claude.agent.sdk.types

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Permission modes for tool usage.
 */
@Serializable
enum class PermissionMode {
    @SerialName("default") DEFAULT,
    @SerialName("acceptEdits") ACCEPT_EDITS,
    @SerialName("plan") PLAN,
    @SerialName("bypassPermissions") BYPASS_PERMISSIONS
}

/**
 * Permission behavior types.
 */
@Serializable
enum class PermissionBehavior(val value: String) {
    @SerialName("allow") ALLOW("allow"),
    @SerialName("deny") DENY("deny"),
    @SerialName("ask") ASK("ask")
}

/**
 * Permission update destination.
 */
@Serializable
enum class PermissionUpdateDestination {
    @SerialName("userSettings") USER_SETTINGS,
    @SerialName("projectSettings") PROJECT_SETTINGS,
    @SerialName("localSettings") LOCAL_SETTINGS,
    @SerialName("session") SESSION
}

/**
 * Permission rule value.
 */
@Serializable
data class PermissionRuleValue(
    val toolName: String,
    val ruleContent: String? = null
)

/**
 * Permission update configuration.
 */
@Serializable
data class PermissionUpdate(
    val type: PermissionUpdateType,
    val rules: List<PermissionRuleValue>? = null,
    val behavior: PermissionBehavior? = null,
    val mode: PermissionMode? = null,
    val directories: List<String>? = null,
    val destination: PermissionUpdateDestination? = null
)

/**
 * Permission update types.
 */
@Serializable
enum class PermissionUpdateType {
    @SerialName("addRules") ADD_RULES,
    @SerialName("replaceRules") REPLACE_RULES,
    @SerialName("removeRules") REMOVE_RULES,
    @SerialName("setMode") SET_MODE,
    @SerialName("addDirectories") ADD_DIRECTORIES,
    @SerialName("removeDirectories") REMOVE_DIRECTORIES
}

/**
 * Context information for tool permission callbacks.
 */
data class ToolPermissionContext(
    @kotlinx.serialization.Contextual
    val signal: Any? = null, // Future: abort signal support
    val suggestions: List<PermissionUpdate> = emptyList() // Permission suggestions from CLI
)

/**
 * Union type for permission results.
 */
sealed interface PermissionResult {
    val behavior: PermissionBehavior
}

/**
 * Allow permission result.
 */
@Serializable
data class PermissionResultAllow(
    override val behavior: PermissionBehavior = PermissionBehavior.ALLOW,
    val updatedInput: Map<String, JsonElement>? = null,
    val updatedPermissions: List<PermissionUpdate>? = null
) : PermissionResult

/**
 * Deny permission result.
 */
@Serializable
data class PermissionResultDeny(
    override val behavior: PermissionBehavior = PermissionBehavior.DENY,
    val message: String = "",
    val interrupt: Boolean = false
) : PermissionResult

/**
 * Tool permission callback function type.
 * @param toolName 工具名称
 * @param input 工具输入参数（JSON 对象）
 * @param toolUseId 工具调用 ID（用于精确关联 UI）
 * @param context 权限上下文
 */
typealias CanUseTool = suspend (toolName: String, input: Map<String, JsonElement>, toolUseId: String?, context: ToolPermissionContext) -> PermissionResult

File: vscode-extension/src/sdk/claude/types/permissions.ts
/**
 * Permission types for Claude Agent SDK
 * Translated from Kotlin: claude-agent-sdk/src/main/kotlin/com/asakii/claude/agent/sdk/types/Permissions.kt
 */

import type { JsonValue, JsonObject } from './common';

/**
 * Permission modes for tool usage.
 */
export type PermissionMode = 'default' | 'acceptEdits' | 'plan' | 'bypassPermissions';

/**
 * Permission mode enum values.
 */
export const PermissionModes = {
  DEFAULT: 'default',
  ACCEPT_EDITS: 'acceptEdits',
  PLAN: 'plan',
  BYPASS_PERMISSIONS: 'bypassPermissions',
} as const;

/**
 * Permission behavior types.
 */
export type PermissionBehavior = 'allow' | 'deny' | 'ask';

/**
 * Permission behavior enum values.
 */
export const PermissionBehaviors = {
  ALLOW: 'allow',
  DENY: 'deny',
  ASK: 'ask',
} as const;

/**
 * Permission update destination.
 */
export type PermissionUpdateDestination =
  | 'userSettings'
  | 'projectSettings'
  | 'localSettings'
  | 'session';

/**
 * Permission update destination enum values.
 */
export const PermissionUpdateDestinations = {
  USER_SETTINGS: 'userSettings',
  PROJECT_SETTINGS: 'projectSettings',
  LOCAL_SETTINGS: 'localSettings',
  SESSION: 'session',
} as const;

/**
 * Permission rule value.
 */
export interface PermissionRuleValue {
  toolName: string;
  ruleContent?: string;
}

/**
 * Permission update types.
 */
export type PermissionUpdateType =
  | 'addRules'
  | 'replaceRules'
  | 'removeRules'
  | 'setMode'
  | 'addDirectories'
  | 'removeDirectories';

/**
 * Permission update type enum values.
 */
export const PermissionUpdateTypes = {
  ADD_RULES: 'addRules',
  REPLACE_RULES: 'replaceRules',
  REMOVE_RULES: 'removeRules',
  SET_MODE: 'setMode',
  ADD_DIRECTORIES: 'addDirectories',
  REMOVE_DIRECTORIES: 'removeDirectories',
} as const;

/**
 * Permission update configuration.
 */
export interface PermissionUpdate {
  type: PermissionUpdateType;
  rules?: PermissionRuleValue[];
  behavior?: PermissionBehavior;
  mode?: PermissionMode;
  directories?: string[];
  destination?: PermissionUpdateDestination;
}

/**
 * Context information for tool permission callbacks.
 */
export interface ToolPermissionContext {
  /** Future: abort signal support */
  signal?: unknown;
  /** Permission suggestions from CLI */
  suggestions?: PermissionUpdate[];
}

/**
 * Union type for permission results.
 */
export type PermissionResult = PermissionResultAllow | PermissionResultDeny;

/**
 * Allow permission result.
 */
export interface PermissionResultAllow {
  behavior: 'allow';
  updatedInput?: Record<string, JsonValue>;
  updatedPermissions?: PermissionUpdate[];
}

/**
 * Deny permission result.
 */
export interface PermissionResultDeny {
  behavior: 'deny';
  message?: string;
  interrupt?: boolean;
}

/**
 * Helper functions to create permission results.
 */
export const PermissionResult = {
  allow(
    updatedInput?: Record<string, JsonValue>,
    updatedPermissions?: PermissionUpdate[]
  ): PermissionResultAllow {
    return { behavior: 'allow', updatedInput, updatedPermissions };
  },
  deny(message?: string, interrupt?: boolean): PermissionResultDeny {
    return { behavior: 'deny', message, interrupt };
  },
};

/**
 * Tool permission callback function type.
 * @param toolName Tool name
 * @param input Tool input parameters (JSON object)
 * @param toolUseId Tool call ID (for precise UI association)
 * @param context Permission context
 */
export type CanUseTool = (
  toolName: string,
  input: Record<string, JsonValue>,
  toolUseId: string | undefined,
  context: ToolPermissionContext
) => Promise<PermissionResult>;
